pkg/operator/deploymentcontroller/deployment_controller.go

```diff
@@ -70,36 +78,111 @@ func NewDeploymentController(
+		deployInformer,
+	).WithConditions(
+		opv1.OperatorStatusTypeAvailable,
+		opv1.OperatorStatusTypeProgressing,
+		opv1.OperatorStatusTypeDegraded,
+	).WithExtraInformers(
+		optionalInformers...,
+	).WithManifestHooks(
+		optionalManifestHooks...,
+	).WithDeploymentHooks(
+		optionalDeploymentHooks...,
+	)
+	return c.ToController()
+}
+
+// NewDeploymentControllerBuilder initializes and returns a pointer to a
+// minimal DeploymentController.
+func NewDeploymentControllerBuilder(
+	name string,
+	manifest []byte,
+	recorder events.Recorder,
+	operatorClient v1helpers.OperatorClientWithFinalizers,
+	kubeClient kubernetes.Interface,
+	deployInformer appsinformersv1.DeploymentInformer,
+) *DeploymentController {
+	return &DeploymentController{
+		name:           name,
+		manifest:       manifest,
+		operatorClient: operatorClient,
+		kubeClient:     kubeClient,
+		deployInformer: deployInformer,
+		recorder:       recorder,
 	}
+}
 
-	informers := append(
-		optionalInformers,
-		operatorClient.Informer(),
-		deployInformer.Informer(),
+// WithExtraInformers appends additional informers to the DeploymentController.
+// These informers are used to watch for additional resources that might affect the Deployment's state.
+func (c *DeploymentController) WithExtraInformers(informers ...factory.Informer) *DeploymentController {
+	c.optionalInformers = informers
+	return c
+}
+
+// WithManifestHooks adds custom hook functions that are called during the handling of the Deployment manifest.
+// These hooks can manipulate the manifest or perform specific checks before its convertion into a Deployment object.
+func (c *DeploymentController) WithManifestHooks(hooks ...ManifestHookFunc) *DeploymentController {
+	c.optionalManifestHooks = hooks
+	return c
+}
+
+// WithDeploymentHooks adds custom hook functions that are called during the sync.
+// These hooks can perform operations or modifications at specific points in the Deployment.
+func (c *DeploymentController) WithDeploymentHooks(hooks ...DeploymentHookFunc) *DeploymentController {
+	c.optionalDeploymentHooks = hooks
+	return c
+}
+
+// WithConditions sets the operational conditions under which the DeploymentController will operate.
+// Only 'Available', 'Progressing' and 'Degraded' are valid conditions; other values are ignored.
+func (c *DeploymentController) WithConditions(conditions ...string) *DeploymentController {
+	validConditions := sets.New[string]()
+	validConditions.Insert(
+		opv1.OperatorStatusTypeAvailable,
+		opv1.OperatorStatusTypeProgressing,
+		opv1.OperatorStatusTypeDegraded,
 	)
+	for _, condition := range conditions {
+		if validConditions.Has(condition) {
+			if !slices.Contains(c.conditions, condition) {
+				c.conditions = append(c.conditions, condition)
+			}
+		}
+	}
+	return c
+}
 
-	return factory.New().WithInformers(
+// ToController converts the DeploymentController into a factory.Controller.
+func (c *DeploymentController) ToController() factory.Controller {
```

Comment:
This could return an error so people would knwo they broke it.

Reply:
Nice! Done.